src/main/java/com/amazon/aws/iot/greengrass/semver/Range.java

```diff
@@ -0,0 +1,486 @@
+            return comp;
+        }
+
+        static boolean isX(String id) {
+            return id == null || id.trim().isEmpty() || id.equalsIgnoreCase("x") || id.equals("*");
+        }
+
+        static String replaceXRanges(String comp) {
+            return Arrays.stream(S_PLUS.split(comp.trim(), -1))
+                    .map(FreeFunctions::replaceXRange)
+                    .collect(Collectors.joining(" "));
+        }
+
+        private static String replaceXRange(String s) {
+            Pattern r = SemVerRegexes.re.get(SemVerRegexes.t.get("XRANGE"));
+            Matcher mm = r.matcher(s);
+            if (!mm.find()) {
+                return s;
+            }
+
+            String gtlt = mm.group(1);
+            String M = mm.group(2);
+            String m = mm.group(3);
+            String p = mm.group(4);
+            String pr = mm.group(5);
+
+            String ret = s;
+            boolean xM = isX(M);
+            boolean xm = xM || isX(m);
+            boolean xp = xm || isX(p);
+            boolean anyX = xp;
+
+            if ("=".equals(gtlt) && anyX) {
+                gtlt = "";
+            }
+
+            pr = "";
+
+            if (xM) {
+                if (">".equals(gtlt) || "<".equals(gtlt)) {
+                    // nothing is allowed
+                    ret = "<0.0.0-0";
+                } else {
+                    // nothing is forbidden
+                    ret = "*";
+                }
+            } else if (!gtlt.isEmpty() && anyX) {
+                // we know patch is an x, because we have any x at all.
+                // replace X with 0
+                if (xm) {
+                    m = "0";
+                }
+                p = "0";
+
+                if (">".equals(gtlt)) {
+                    // >1 => >=2.0.0
+                    // >1.2 => >=1.3.0
+                    gtlt = ">=";
+                    if (xm) {
+                        M = String.valueOf(Integer.parseInt(M) + 1);
+                        m = "0";
+                        p = "0";
+                    } else {
+                        m = String.valueOf(Integer.parseInt(m) + 1);
+                        p = "0";
+                    }
+                } else if ("<=".equals(gtlt)) {
+                    // <=0.7.x is actually <0.8.0, since any 0.7.x should
+                    // pass.  Similarly, <=7.x is actually <8.0.0, etc.
+                    gtlt = "<";
+                    if (xm) {
+                        M = String.valueOf(Integer.parseInt(M) + 1);
+                    } else {
+                        m = String.valueOf(Integer.parseInt(m) + 1);
+                    }
+                }
+
+                if ("<".equals(gtlt)) {
+                    pr = "-0";
+                }
+
+                ret = String.format("%s%s.%s.%s%s", gtlt, M, m, p, pr);
+            } else if (xm) {
+                ret = String.format(">=%s.0.0%s <%s.0.0-0", M, pr, Integer.parseInt(M) + 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;+&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 292. Consider formatting this data with &#39;%d&#39; instead.